File: opinionpulse/opinionpulse-backend/app/schemas/market.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel


class PricePoint(BaseModel):
    time: str
    price: float


class MarketChartResponse(BaseModel):
    query: str
    asset_type: Literal["crypto", "stock", "unknown"]
    symbol: Optional[str] = None
    name: str
    current_price: Optional[float] = None
    change_pct: Optional[float] = None
    currency: str = "USD"
    points: list[PricePoint]
    message: Optional[str] = None

File: opinionpulse/opinionpulse-frontend/src/api/market.ts
import { apiRequest } from "@/api/client"

export type MarketChartPoint = {
  time: string
  price: number
}

export type MarketChartResponse = {
  query: string
  asset_type: "crypto" | "stock" | "unknown"
  symbol: string | null
  name: string
  current_price: number | null
  change_pct: number | null
  currency: string
  points: MarketChartPoint[]
  message: string | null
}

export async function getMarketChart(query: string): Promise<MarketChartResponse> {
  const params = new URLSearchParams({ q: query })
  return apiRequest<MarketChartResponse>(`/api/market/chart?${params.toString()}`, {
    auth: true,
  })
}
